hedyorg/hedy
# Autocomplete usernames when inviting students and teachers

Adds autocomplete functionality for searching usernames when adding a student or a second teacher.

It does so by matching the usernames in the database with the user's query, so if search for: user, it'll show me: user1, user2, user3, etc.

In the process of doing this I wanted to use a bit more of the power of Dynamo, instead of relying on filtering the data trhough python. To this end I added 2 extra Dynamo conditions: one for checking if a set exits, and another for checking if an element is not in a container.

I also updated HTMX and Tailwind, but maybe I can update those in a different PR?

I tried to improve the code a little bit also by avoiding duplication where I could, so instead of having 2 function for adding teachers and students I merged them together since they were mostly the same.

Fixes #4689  

**How to test**

1. Log is as a teacher.
2. Go to the class page
3. Click on Add student and then on invite by username
4. Search and check that a) no students that are already in the class show up.
5. Select a few and send invitations. Check that they were sent
6. Now click on Invite second teacher. Check that: a) only users that are teachers are shown, b) only teachers that aren't already second teacher's in the class show.
7. Send invitations and see if they are added correctly.
